nservicebus/azure-service-bus/addressing-logic.md

```diff
@@ -0,0 +1,174 @@
+
+Out of the box there are 3 namespace partitioning strategies
+ * `SingleNamespacePartitioning` (default): All entities are in a single namespace.
+ * `FailOverNamespacePartitioning`: assumes all entities are in the primary and secondary namespaces, where only the primary is in use by default. The secondary will function as a fallback in case of problems with the primary.
+ * `RoundRobinNamespacePartitioning`: assumes all entities are in all namespaces, all namespaces are in use but one at a time in a round robin fashion.
+
+The default implementation of this strategy can be replaced by using the configuration API:
+
+snippet:swap-namespace-partitioning-strategy
+
+#### Implementing a custom namespace partitioning strategy
+
+Implementing a custom namespace partitioning strategy requires a class that implements `INamespacePartitioningStrategy`. This interface contains one method, called `GetNamespaces`, which provides access to the `PartitioningIntent` for which namespaces are requested and returns a set of namespaces, represented by the `RuntimeNamespaceInfo` class, that should be used for the specified intent.
+
+The `PartitioningIntent` is to be interpreted as follows:
+* `Creating`: returns a set of namespaces in which entities should be created at startup, most likely all namespaces.
+* `Receiving`: returns a set of namespaces to which the transport should listen, very likely to be all namespaces as well.
+* `Sending`: returns a set of namespaces to which the next send operation should go. In this operation it's very important to differentiate between `NamespaceMode.Active` and `NamespaceMode.Passive`. The transport will send the message to **ALL** active namespaces and if one of those send operation fails it will attempt to execute that operation on one of the passive namespaces returned.
+
+If the implementation of a namespace partitioning strategy requires configuration settings, these settings can be accessed by letting NServiceBus inject the `ReadOnlySettings` into the constructor of the strategy.
+
+snippet:custom-namespace-partitioning-strategy
+
+#### Extending the configuration API for custom partitioning settings
+
+In order to allow configuration of the custom namespace partitioning strategy it is advised to create an extension method at the `AzureServiceBusNamespacePartitioningSettings` extension point , returned by `.NamespacePartitioning()`,in the NServiceBus configuration API. This provides access to the settings, in which the value can be registered using a well known key.
+
+snippet:custom-namespace-partitioning-strategy-extension
+
+### Composition
+
+The composition aspect is represented by an implementation of `ICompositionStrategy`.
+
+Out of the box there are 2 composition strategies
+ * `FlatComposition`: The entity is in the root of the namespace.
+ * `HierarchyComposition`: The entity is in a namespace hierarchy, at the location generated by the path generator.
```

Comment:
Currently this is broken (https://github.com/Particular/NServiceBus.AzureServiceBus/issues/190) as there's no way to supply path generator for the strategy.
PR https://github.com/Particular/NServiceBus.AzureServiceBus/pull/225 will address it and add a configuration API to supply path generator.